### ISSUE TITLE
add support for other devices - Avidyne ?

### ISSUE THREAD

**Issue by n251ea** [dimaryaz/jdmtool#6]
Maybe add support for other devices such as the Avidyne IFD navigator?  JDM downloads 3 files
navdata.dsf - the navdata
obstacles.dsf - the obstacle database
.jdm - looks to be a simple text file except 3 numbers change whenever the files are downloaded, I suspect they are either encoded time and/or location on the thumb drive.  If you copy all 3 files to another thumbdrive they download will not instal

JDM requires any thumbdrive that is FAT32

jdm file contents (some items masked for privacy) for navdata only -

{"ss":[{"a":"Avidyne IFD 400 Series, Bendix King AeroNav Series","c":"XXXXX","cd":"North America","date_label_override":"Effective Date:","dv":"2404","f":[{"fh":"eacbcd26
","fp":"navdata.dsf","fs":15412018,"sh":"33da0052"}],"fid":"","filter_applied":"no","gsi":"","jvsn":"","ndv":"2405","nvad":"2024-05-06 06:00:00","nvsd":"2024-05-16 06:00:
00","oatn":"YYYYY","pi":"DAE40253","s":"NavData","sc":"North America","u":"42029628","uv":"42029628_2404","v":"2404","ved":"2024-05-16 06:00:00","vsd":"2024-04-18 06:00:
00"}],"ver":"1.1","z":"0368217f"}

XXXXX - is the jeppesen customer number
YYYYY - is the aircraft N number

fp - filename to load
fs - file size in bytes
sh - unknown and changes for each time the (same) database is downloaded 
pi - unknown but doesn't change when downloading the same database
u & uv - unknown but doesn't change when downloading the same database
z - unknown and changes for each time the (same) database is downloaded

The obstacle file entry is similar, an actual file can be provided

I suspect the numbers that change each time the same database is downloaded are either time/date encodings or location/sectors on the thumbdrive as you cannot simply use "cp" to move all the files to a new thumbdrive.

**Comment by dimaryaz** [comment]
It might be doable, depending on how much JDM obfuscates the data.

Does `jdmtool` download anything successfully - at least the .dsf files? Would you mind posting your services.xml? (Should be in `~/.local/share/jdmtool/services.xml`; you should mask anything sensitive - your username, email, name, customer ID, N number). I may also need an example of the downloaded databases - but we'll see; their server may let me download them even with the wrong subscription.

> I suspect the numbers that change each time the same database is downloaded are either time/date encodings or location/sectors on the thumbdrive as you cannot simply use "cp" to move all the files to a new thumbdrive.

Yeah, that's completely possible. I've been trying to reverse engineer G1000 for a while; it uses the SD card serial number, and [encodes it to hide it](https://github.com/dimaryaz/jdmtool/blob/g1000/src/jdmtool/gplugin.py#L172-L173).

**Comment by n251ea** [comment]
I can confirm that the .jdm file doesn't have a checksum in it and if I simply change the N number in the .jdm file you can load the files into a different unit, and if you download the subscription for two different users/aircraft the datafiles are exactly the same so everything is controlled by the .jdm file.

here's the output from jdmtool,  it downloads my database as a zip file

me@laptop:~$ jdmtool login
Username: me@somewhere.org
Password:
Logged in successfully
me@laptop:~$ jdmtool refresh
Downloading services...
Downloading keychain...
Success
me@laptop:~$ jdmtool list
ID  Name                                                                    Coverage              Version   Start Date  End Date    Downloaded
 0  Avidyne IFD 400 Series, Bendix King AeroNav Series - NavData            North America         2404      2024-04-18  2024-05-16
me@laptop:~$ jdmtool download 0 
Downloading database: 100%|████████████████| 21.7M/21.7M [00:05<00:00, 4.17MB/s]
Downloaded to /home/me/.local/share/jdmtool/downloads/dae402_2404_bbbb36dd.zip
me@laptop:~$ 

<?xml version="1.0" encoding="UTF-8"?>
<getserviceslist>


<username>me@somewhere.org</username>
<datetime>2024-05-04 12:53:38</datetime>
<page_generation_date>1714827218</page_generation_date>
<server_name>jdm-fbc9d47d5-2jbg8</server_name>
<customer_id>XXXXX</customer_id>
<customer_num>YYYYY</customer_num>
<email>me@somewhere.org</email>
<customer_name>just me</customer_name>
<service><customer_number>YYYYY</customer_number><category>10</category><service_code>DAE40253</service_code><order_code>DAE402</order_code><unique_service_id>42029628</unique_service_id><product_item>DAE40253</product_item><reference_number>42029628</reference_number><oracle_aircraft_tail_number>N12345</oracle_aircraft_tail_number><oracle_aircraft_manufacturer>PIPER</oracle_aircraft_manufacturer><oracle_aircraft_model>PA23, AZTEC, APACHE</oracle_aircraft_model><service_type>NavData</service_type><service_type_code>NavData</service_type_code><avionics>Avidyne IFD 400 Series, Bendix King AeroNav Series</avionics><coverage_desc>North America</coverage_desc><oem_avidyne_e2>1</oem_avidyne_e2><oem_metadata>1</oem_metadata><data_revision>0</data_revision><filename>dae402_2404_bbbb36dd.zip</filename><version>2404</version><display_version>2404</display_version><max_initial_counter>1</max_initial_counter><downloads>0</downloads><file_size>21672165</file_size><file_crc>bbbb36dd</file_crc><downloads_type>TRUE</downloads_type><short_desc>NavData Avidyne IFD 400 Series North America</short_desc><long_desc>NavData Avidyne IFD 400 Series North America</long_desc><service_renewal_date>2024-12-28 00:00:00</service_renewal_date>^M
<version_start_date>2024-04-18 06:00:00</version_start_date>^M
<version_end_date>2024-05-16 06:00:00</version_end_date>^M
<next_version>2405</next_version>^M
<next_display_version>2405</next_display_version>^M
<next_version_avail_date>2024-05-06 06:00:00</next_version_avail_date>^M
<next_version_start_date>2024-05-16 06:00:00</next_version_start_date>^M
<date_label_override>Effective Date:</date_label_override>^M
<media_image>images/media/jepp_thumb_drive.jpg</media_image><media_image_crc>0x39D88DFA</media_image_crc><coverage_image>images/coverages/n_america.png</coverage_image><coverage_image_crc>0xDF3995CB</coverage_image_crc><media><card_type>2</card_type>^M
<install_dir>\\</install_dir>^M
<filename>charts.dsf</filename>^M
<card_format>2</card_format>^M
<required>0</required>^M
</media>^M
<media><card_type>2</card_type>^M
<install_dir>\\</install_dir>^M
<filename>navdata.dsf</filename>^M
<card_format>2</card_format>^M
<required>0</required>^M
</media>^M
<media><card_type>2</card_type>^M
<card_format>2</card_format>^M
<required>1</required>^M
</media>^M
</service>^M
<response_code>0x0</response_code>
<gen_time>6.3</gen_time></getserviceslist>

**Comment by dimaryaz** [comment]
Ok nice, the XML file worked - I've successfully downloaded the DB. Though it's a .zip file containing a whole bunch of files - but not navdata.dsf or obstacles.dsf - so figuring out where those come from could be a challenge, too. Anyways, I'll try the real JDM next, and see what happens.

**Comment by n251ea** [comment]
I logged in with a different account and found that the obstacle database down downloadable with "jdmtool download 1"  I have attached the xml file for that one as well.  Also I can provide the bin file that the Jeppesen tool makes for this cycle, I suspect they are simply joining the files together to make a single bin file.  I believe the avidyle is using some sort of unix OS so it might be as simple as making a tar or cpio archive with a simple offset.
[services.txt](https://github.com/dimaryaz/jdmtool/files/15213435/services.txt)
